humza1400/apollo
# Fix flaky

# What is the purpose of this PR?
The main goal of this PR is to address and fix a flaky test found in the MiddlewaresTest.java. The test asShouldSerializeObjectAsJson intermittently fails due to inconsistencies in JSON parsing and validation.

# Why the test fails
The test fails intermittently because it attempts to parse the JSON string and extract integer and string values directly. This approach can lead to failures depending on the ordering and formatting of the JSON content. The process of extracting values from the JSON string wasn't robust enough to handle all scenarios and edge cases.

# How to reproduce the test failure
1. Clone the repository and checkout to the commit a03105ee1f98f8c5878f4e3c4d03c8732c6384bd.
2. Run the test suite multiple times or use a tool like NonDex to introduce randomness and observe the intermittent failures.
> mvn -pl apollo-api-impl edu.illinois:nondex-maven-plugin:2.1.1:nondex -Dtest=com.spotify.apollo.route.MiddlewaresTest#asShouldSerializeObjectAsJson

# Expected results
The test is expected to consistently pass, regardless of the ordering or formatting of the JSON string. The values extracted from the JSON should match the expected integer and string values.

# Actual results
The test intermittently fails, returning an assertion error. This inconsistency is due to the non-deterministic nature of the JSON string's ordering and content. This leads to scenarios where the extracted values are not as expected

# Description of fix
We introduced a more reliable method of parsing and validating the JSON string. We leveraged Jackson, a JSON processing library already included in the project dependencies, to deserialize the JSON string into a Java object. This approach ensures that the extraction of values is done in a more type-safe and reliable manner, preventing failures related to JSON string inconsistencies.
